File: 0015-3sum/0015-3sum.js
var threeSum = function (nums) {
    if (!nums) return []
    let result = []
    nums.sort((a, b) => a - b)
    let i = 0

    while (i < nums.length - 2) {

        let left = i + 1
        let right = nums.length - 1

        while (left < right) {
            let currSum = nums[i] + nums[left] + nums[right] // -1

            if (currSum === 0) {
                result.push([nums[i], nums[left], nums[right]])
                if (nums[i] === nums[i + 1]) {
                    while (nums[i] === nums[i + 1]) {
                        i++
                    }
                }

                while (left < right && nums[left] === nums[left + 1]) {
                        left++
                    }
                while (left < right && nums[right] === nums[right - 1]) {
                    right--
                }
                left++
                right--
            } else if (currSum < 0) {
                if (nums[left] === nums[left + 1]) {
                    while (left < right && nums[left] === nums[left + 1]) {
                        left++
                    }
                } else {
                    left++
                }
                // left++
            } else {
                if (nums[right] === nums[right - 1]) {
                    while (left < right && nums[right] === nums[right - 1]) {
                        right--
                    }
                } else {
                    right--
                }
                // right--
            }
        }
        i++
    }

    return result
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        triplets = set()
        nums.sort()
        n = len(nums)

        for i in range(0, n - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            left, right = i + 1, n - 1

            while left < right:
                total = nums[i] + nums[left] + nums[right]
                if total == 0:
                    triplets.add((nums[i], nums[left], nums[right]))
                    left += 1
                    right -= 1
                elif total < 0:
                    left += 1
                else:
                    right -= 1
            
        return [list(triple) for triple in triplets]


